### PR TITLE
[js/rn] Implement blob exchange by JSI instead of use base64

### PR DESCRIPTION
**Pull request by jhen0409**
### Description

- Create `OnnxruntimeJSIHelper` native module to provide two JSI functions
  - `jsiOnnxruntimeStoreArrayBuffer`: Store buffer in Blob Manager & return blob object (iOS: RCTBlobManager, Android: BlobModule)
  - `jsiOnnxruntimeResolveArrayBuffer`: Use blob object to get buffer
  - The part of implementation is reference to [react-native-blob-jsi-helper](https://github.com/mrousavy/react-native-blob-jsi-helper)
- Replace base64 encode/decode
  - `loadModelFromBlob`: Rename from `loadModelFromBase64EncodedBuffer`
  - `run`: Use blob object to replace input.data & results[].data

For [this context](https://github.com/microsoft/onnxruntime/issues/16031#issuecomment-1556527812), it saved a lot of time and avoid JS thread blocking in decode return type, it is 3700ms -> 5~20ms for the case. (resolve function only takes 0.x ms)

### Motivation and Context

It’s related to #16031, but not a full implementation for migrate to JSI.

It just uses JSI through BlobManager to replace the slow part (base64 encode / decode).

Rewriting it entirely in JSI could be complicated, like type convertion and threading. This PR might be considered a minor change.

/cc @skottmckay

### CODE REVIEW

**Comment by jhen0409** [comment]
@microsoft-github-policy-service agree

**Comment by fs-eire** [comment]
The typescript part looks good to me.

I don't know much details about JSI. I have 2 questions regarding the implementation:
- is there a way to export `jsiOnnxruntimeStoreArrayBuffer` and `jsiOnnxruntimeResolveArrayBuffer` to the module object instead of `global`? In general exporting to global is not preferred.
- is the lifecycle of the allocated `ArrayBuffer`s correctly managed? I want to make sure that buffers are not released too early (otherwise a UB in memory may happen) or too late/never (a memory leak)

**Comment by jhen0409** [comment]
> * is there a way to export `jsiOnnxruntimeStoreArrayBuffer` and `jsiOnnxruntimeResolveArrayBuffer` to the module object instead of `global`? In general exporting to global is not preferred.

It probably can't, because the install function is use `RCT_EXPORT_BLOCKING_SYNCHRONOUS_METHOD` not a JSI function, so we can't use JSI object as argument or return type.

The official way is use [Turbo Native Module](https://reactnative.dev/docs/the-new-architecture/cxx-cxxturbomodules), but it required new architecture enabled and currently not stable.

We can just delete it in global after installation, but not perfect.

> * is the lifecycle of the allocated ArrayBuffers correctly managed? I want to make sure that buffers are not released too early (otherwise a UB in memory may happen) or too late/never (a memory leak)

Currently the implementation is take the `ArrayBuffer` constructor from JS, so it should be managed by the JS runtime. Also it remove blob in blobManager after it becomes ArrayBuffer.

For iOS, the blob (NSData*) should dealloc by ARC. For Android, the bytes reference (jbyte*) should managed by Java.

**Comment by jhen0409** [comment]
> **_UPDATE:_** I found some memory leak issue on load model with my examples, not sure if it is related to this PR (it seems fine by patch this to v1.14), will investigate this.

Confirmed it is not related to this PR, looks like some load model behaviors are changed from v1.14, and it's bug I think. I'll open another issue or PR.

**Review by fs-eire — APPROVED** [review]
I think it is good for the typescript part and my concerns are resolved. Please wait for @skottmckay review the android/ios part.

File: js/react_native/android/src/main/cpp/cpp-adapter.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    env->DeleteLocalRef(jstring);
+
+    jboolean isCopy = true;
+    jbyte* bytes = env->GetByteArrayElements(boxedBytes, &isCopy);
+    env->DeleteLocalRef(boxedBytes);
+    return reinterpret_cast<byte*>(bytes);
+  };
+
+  auto resolveArrayBuffer = jsi::Function::createFromHostFunction(runtime,
+                                                                  jsi::PropNameID::forAscii(runtime, "getArrayBufferForBlob"),
```
Comment (YUNQIUGUO) [inline js/react_native/android/src/main/cpp/cpp-adapter.cpp:60]: > "getArrayBufferForBlob",

nit: looks like the function is called `resolveArrayBuffer`,  should we update this name here?

File: js/react_native/android/src/main/java/ai/onnxruntime/reactnative/OnnxruntimeJSIHelper.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Override
+  @NonNull
+  public String getName() {
+    return NAME;
+  }
+
+  public void checkBlobModule() {
+    if (blobModule == null) {
+      blobModule = getReactApplicationContext().getNativeModule(BlobModule.class);
```
Comment (YUNQIUGUO) [inline js/react_native/android/src/main/java/ai/onnxruntime/reactnative/OnnxruntimeJSIHelper.java:35]: should we throw exceptions if not able to get  `BlobModule` ?

File: js/react_native/ios/OnnxruntimeModuleTest/TensorHelperTest.mm [hunk 3]
```diff
@@ -208,9 +217,26 @@ static void testCreateOutputTensorT(const std::array<T, 5> &outValues, std::func
...
 
-  NSDictionary *resultMap = [TensorHelper createOutputTensor:outputNames values:output];
+  NSDictionary *resultMap = [TensorHelper createOutputTensor:testBlobManager outputNames:outputNames values:output];
+
+  // Compare output & input, but data.blobId is different
+
+  NSDictionary *outputMap = [resultMap objectForKey:@"output"];
+
+  // Log outputMap[@"dims"]
+  NSLog(@"outputMap[dims]: %@", outputMap[@"dims"]);
```
Comment (YUNQIUGUO) [inline js/react_native/ios/OnnxruntimeModuleTest/TensorHelperTest.mm:227]: nit: Log here

**Review by YUNQIUGUO — COMMENTED** [review]
thanks for the contribution!

File: js/react_native/android/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+set (PACKAGE_NAME "onnxruntime-react-native")
+set (BUILD_DIR ${CMAKE_SOURCE_DIR}/build)
+set(CMAKE_VERBOSE_MAKEFILE ON)
+set(CMAKE_CXX_STANDARD 17)
+
+file(TO_CMAKE_PATH "${NODE_MODULES_DIR}/react-native/ReactCommon/jsi/jsi/jsi.cpp" libPath)
+
+include_directories(
+        ../cpp
```
[the remark this answers is not in the document]
Comment (jhen0409) [reply js/react_native/android/CMakeLists.txt:12]: This is not used (before I was going to make a ./cpp utils), I'll remove it.

File: js/react_native/android/src/main/cpp/cpp-adapter.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  env->ReleaseByteArrayElements(stringJbytes, pBytes, JNI_ABORT);
+
+  env->DeleteLocalRef(stringJbytes);
+  env->DeleteLocalRef(stringClass);
+  return ret;
+}
+
+extern "C"
+JNIEXPORT void JNICALL
+Java_ai_onnxruntime_reactnative_OnnxruntimeJSIHelper_nativeInstall(JNIEnv *env, jclass _, jlong jsiPtr, jobject instance) {
```
[the remark this answers is not in the document]
Comment (jhen0409) [reply js/react_native/android/src/main/cpp/cpp-adapter.cpp:30]: The getBytesFromBlob / createBlob function is used JNI env & instanceGlobal refs from nativeInstall. It's easily refactored to function like jstring2string.

File: js/react_native/lib/binding.ts [hunk 6]
```diff
@@ -64,11 +73,34 @@ export declare namespace Binding {
 
   interface InferenceSession {
     loadModel(modelPath: string, options: SessionOptions): Promise<ModelLoadInfoType>;
-    loadModelFromBase64EncodedBuffer?(buffer: string, options: SessionOptions): Promise<ModelLoadInfoType>;
+    loadModelFromBlob?(buffer: string|JSIBlob, options: SessionOptions): Promise<ModelLoadInfoType>;
```
[the remark this answers is not in the document]
Comment (jhen0409) [reply js/react_native/lib/binding.ts:74]: It's a mistake, I'll fix it.

File: js/react_native/lib/binding.ts [hunk 7]
```diff
@@ -64,12 +71,41 @@ export declare namespace Binding {
...
 export const binding = Onnxruntime as Binding.InferenceSession;
+
+// install JSI helper global functions
+OnnxruntimeJSIHelper.install();
+
+declare global {
+  // eslint-disable-next-line no-var
+  var jsiOnnxruntimeStoreArrayBuffer: ((buffer: ArrayBuffer) => JSIBlob)|undefined;
+  // eslint-disable-next-line no-var
+  var jsiOnnxruntimeResolveArrayBuffer: ((blob: JSIBlob) => ArrayBuffer)|undefined;
```
Comment (jhen0409) [inline js/react_native/lib/binding.ts:91]: I disable `no-var` for ESLint because it required to use `var` for declare global in TS.

**Comment by jhen0409** [comment]
```
C:\a\_work\1\s\js\react_native\lib\backend.ts
  126:18  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

Sorry about that, it doesn't happened in my environment, but it's the obvious part that can be corrected so I'll fix it.

**Comment by skottmckay** [comment]
Sorry - the PR to fix the broken CI pipeline needed a second fix. That is now checked in and the pipeline has been passing for the last day. Could you please update to the latest `main` one more time? Apologies for the inconvenience.

**Comment by jhen0409** [comment]
> Sorry - the PR to fix the broken CI pipeline needed a second fix. That is now checked in and the pipeline has been passing for the last day. Could you please update to the latest `main` one more time? Apologies for the inconvenience.

OK, updated.

**Review by skottmckay — APPROVED** [review]

**Comment by YUNQIUGUO** [comment]
Did we manually run the ONNX Runtime React Native CI pipeline on this pr before? I think this change breaks the react native E2E test again...

**Pull request merged**